### ISSUE TITLE
[DDC] Create incremental mode in DDC for evaluation

### ISSUE THREAD

**Issue by annagrin** [dart-lang/sdk#41480]
**Short description**

Make `ProgramCompiler.emitFunctionIncremental` define symbols used in emitted JavaScript.

**Problem**

dartdevc's `ProgramCompiler` is does not currently support incremental compilation, which is needed for expression evaluation. We created workarounds for  incremental function compilation to work, which are good for initial implementation but need to be replaced by proper implementation for stability.

Expression evaluation needs to be able to compile dart expressions to JavaScript snippets that can be evaluated in the browser in scope of the current breakpoint. Currently the JavaScript generated by `ProgramCompiler.emitFunctionIncremental` is missing symbol definitions for symbols used in the evaluated expression.

As a temporary workaround, current implementation of `ExpressionEvaluator` (in dwds) sends supporting information to the `ExpressionCompiler `(in dartdevc), which uses it to create additional statements defining symbols for libraries and fields.  This workaround might break on ddc changes, and does not support all module systems. 

**Example**

```
class Class {
  final int field;
  final int _field;
  Class(this.field, this._field);
}

void main() {
  var instance = Class(1, 2);
  print('$instance'); // ** Breakpoint here **
}
```

To evaluate the expression `instance._field` on `breakpoint here` line, expression compiler produces the following JavaScript code:

```
function(local, instance) {
  let web__main = require('web/main').web__main;
  let _field = dart.privateName(web__main, "_field");
  return instance[_field];
}
```

where only the statement  `return instance[_field];` is actually produced by the ddc. The first two are generated inside the expression compiler, and are needed to re-define symbols used in the produced JavaScript but not necessarily available in the current scope.

**Suggestion:**

The program compiler has all the information needed to define the library and symbols variables. We can make `emitFunctionIncremental` emit  code that can be evaluated in the current scope without workarounds:
 
In general, for every symbol used in the evaluated expression, we need to re-define the symbol's variable and and, transitively, all symbols used in that definition. The symbols might include fields, types, constants, libraries, etc.

After that, we can remove the workaround code and all no longer needed supporting information send from the debugger, to minimize data transfer.

**Follow up with changes in dwds**

https://github.com/dart-lang/webdev/issues/1139

**Related**

https://github.com/dart-lang/sdk/issues/44979
https://github.com/dart-lang/sdk/issues/44713
https://github.com/dart-lang/sdk/issues/44713
https://github.com/dart-lang/sdk/issues/44933
https://github.com/dart-lang/sdk/issues/44813
https://github.com/dart-lang/sdk/issues/44686

**Comment by annagrin** [comment]
replaced by https://github.com/dart-lang/webdev/issues/1139 with better explanations.

**Comment by sigmundch** [comment]
/cc @nshahan, @Markzipan - note that this relates to our general conversation about encapsulating DDC's output.

@annagrin - in general, we have been discussing to get to a place where DDC's output is sealed better. We'd like for everything it emits to be self contained or explicitly exposed using a js API. For expression evaluation, we had discussed that we would attach to each library/module objects exposing what an incremental output would need. That way the output of emitFunctionIncremental can access everything it needs without using an enclosing scope.

Besides expression evaluation, we want to similarly shift to generate the bootstraping logic to the DDC codebase. We've noticed that bootstrappers in flutter tools and webdev are using DDC's output directly in ways that may not be robust or easy to detect for us whether we are breaking an unspoken convention
